### DISCUSSION TITLE
requestAsyncStorage error when using cookies() in middleware

### DISCUSSION THREAD

**Discussion by ricardo-fnd** [vercel/next.js#54853] [Help] [answered]
### Summary

I have an application that has a `middleware`. The objective of that `middleware` is to redirect the user to a certain page depending on the user.
So, I need to fetch some api information about the user, so I can redirect him. When importing `import { cookies } from 'next/headers'` in the `middleware` this error is thrown. 

`Error: Invariant: Method expects to have requestAsyncStorage, none available`

I know, I can use `request.cookies` but the problem is that I'm using a shared service for server components, that uses `cookies` from `next/headers`.
Is this the problem? importing `cookies from next/auth` in the middleware?

Thanks

### Additional information

<img width="729" alt="Screenshot 2023-08-31 at 17 55 50" src="https://github.com/vercel/next.js/assets/52207061/69f14a4d-06c6-40c0-894b-5a2f0930b431">

**Comment by devjiwonchoi** [comment]
I'm assuming that it is because the `middleware` runs before the request is completed, which is failing to read `requestAsyncStorage`.

Apologies for missing out on an assured reason for it. I'll update if I get to know the clear reason.

As far as I know, the description of `cookies` function states that its usage is for server components, route handlers, or server actions.

> The cookies function allows you to read the HTTP incoming request cookies from a [Server Component](https://nextjs.org/docs/app/building-your-application/rendering/server-components) or write outgoing request cookies in a [Server Action](https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations) or [Route Handler](https://nextjs.org/docs/app/building-your-application/routing/route-handlers).

Try to use `(NextRequest | NextResponse).cookies`.

```ts
import { type NextRequest, NextResponse } from 'next/server'

export default function middleware(req: NextRequest) {
  const cookie = req.cookies.get('cookie-name')
  return NextResponse.json({ cookie })
}
```

### Reference

- [Middleware using cookies](https://nextjs.org/docs/app/building-your-application/routing/middleware#using-cookies)

**Reply by ricardo-fnd** [reply]
Using the request object to access cookies property works, but that way, I will have to duplicate services just for the middleware

**Reply by devjiwonchoi** [reply]
I get what you mean, but the `cookies` function is **not available in middleware**, and there is no workaround for this.
Route handlers won't work either since it check from the server component's request only.

**Comment by ztanner** [comment, marked as answer]
Hey @ricardo-fnd , which version of Next are you on? We recently [landed a change](https://github.com/vercel/next.js/pull/53465) to make `cookies()` available in middleware, curious if you’re running into a bug or just haven’t upgraded. It should be available as of 13.4.13.

**Reply by devjiwonchoi** [reply]
Hi @ztanner if you don't mind this is my current state:

### Middleware

```ts
import { NextResponse } from 'next/server'
import { cookies } from 'next/headers'

export default function middleware() {
  const cookie = cookies().get('cookie-name')
  return NextResponse.json({ cookie })
}
```

### Output

![Screenshot 2023-09-01 at 9 14 12 PM](https://github.com/vercel/next.js/assets/120007119/52f1aa0e-c7a0-4933-9bff-b6a62ebdac6d)

### Next Info

```console
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:52 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T8103
    Binaries:
      Node: 18.17.0
      npm: 9.6.7
      Yarn: 1.22.19
      pnpm: 8.6.12
    Relevant Packages:
      next: 13.4.19
      eslint-config-next: 13.4.19
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

**Reply by ztanner** [reply]
Hey @devjiwonchoi, that's odd - can you share a repro? I attempted the same here but am not getting errors:

https://codesandbox.io/p/sandbox/vibrant-tamas-tw2fwm?file=/middleware.ts:9,1

**Reply by ztanner** [reply]
Ah, I think I see what might be happening in your case. Looks like it's not behaving as expected when used with a `src` directory. I'll get a fix up shortly, thanks for flagging

**Reply by devjiwonchoi** [reply]
repro: https://codesandbox.io/p/sandbox/jovial-river-yvqfrm?file=%2Fsrc%2Fmiddleware.ts%3A7%2C2

If you don't mind can I work on it and request a review from you?
If so, would be great if you share which PR it was that allowed middleware to use cookies.

**Reply by ztanner** [reply]
Sorry 🙈 just now noticing this.
I've opened a PR here if you'd like to take a look: https://github.com/vercel/next.js/pull/54889 -- relatively small change to how we're detecting if the invocation is from middleware

**Reply by devjiwonchoi** [reply]
Thank you for sharing!

**Reply by ricardo-fnd** [reply]
I'm also using `/src`. Could you let us know when this goes live @ztanner ? thank you :)

**Reply by armeono** [reply]
has this been fixed? still getting the same error with cookies() from "next-headers".

**Reply by devjiwonchoi** [reply]
@armeono Seems like it's working fine on my condition:

```
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.0.0: Fri Sep 15 14:41:34 PDT 2023; root:xnu-10002.1.13~1/RELEASE_ARM64_T8103
Binaries:
  Node: 18.18.0
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: 8.6.12
Relevant Packages:
  next: 13.5.6-canary.1
  eslint-config-next: 13.5.6-canary.1
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```
